src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java

```diff
@@ -48,193 +47,221 @@
-                final Imp modifiedImp = modifyImp(imp, pubmaticImpExt);
-                modifiedImps.add(modifiedImp);
-                extImpPubmatics.add(extImpPubmatic);
+                if (imp.getBanner() == null && imp.getVideo() == null) {
+                    throw new PreBidException(String.format("Invalid MediaType. PubMatic only supports "
+                            + "Banner and Video. Ignoring ImpID=%s", imp.getId()));
+                }
+                final PubmaticBidderImpExt impExt = parseImpExt(imp);
+                final ExtImpPubmatic extBidder = impExt.getBidder();
+                if (pubId == null) {
+                    pubId = StringUtils.trimToNull(extBidder.getPublisherId());
+                }
+                final ObjectNode extWrapper = impExt.getBidder().getWrapper();
+                if (wrapper == null && extWrapper != null && !extWrapper.isEmpty()) {
+                    validateExtWrapper(extWrapper, imp.getId());
+                    wrapper = extWrapper;
+                }
+
+                validImps.add(modifyImp(imp, impExt));
             } catch (PreBidException e) {
                 errors.add(BidderError.badInput(e.getMessage()));
             }
         }
 
-        if (modifiedImps.isEmpty()) {
+        if (validImps.isEmpty()) {
             return Result.withErrors(errors);
         }
 
-        return Result.of(Collections.singletonList(makeRequest(bidRequest, modifiedImps, extImpPubmatics)), errors);
+        return Result.of(Collections.singletonList(createRequest(request, validImps, pubId, wrapper)), errors);
     }
 
-    private PubmaticImpExt parseImpExt(Imp imp) {
+    private PubmaticBidderImpExt parseImpExt(Imp imp) {
         try {
-            return mapper.mapper().convertValue(imp.getExt(), PubmaticImpExt.class);
+            return mapper.mapper().convertValue(imp.getExt(), PubmaticBidderImpExt.class);
         } catch (IllegalArgumentException e) {
-            throw new PreBidException(e.getMessage(), e);
+            throw new PreBidException(e.getMessage());
         }
     }
 
-    private Imp modifyImp(Imp imp, PubmaticImpExt pubmaticImpExt) throws PreBidException {
-        // validate Impression
-        final ExtImpPubmatic extImpPubmatic = pubmaticImpExt.getBidder();
-        if (imp.getBanner() == null && imp.getVideo() == null) {
-            throw new PreBidException(String.format("Invalid MediaType. PubMatic only supports Banner and Video. "
-                    + "Ignoring ImpID=%s", imp.getId()));
+    private Imp modifyImp(Imp imp, PubmaticBidderImpExt impExt) {
+        final Imp.ImpBuilder impBuilder = imp.toBuilder().audio(null);
+
+        final Banner banner = imp.getBanner();
+
+        updateImpWithAdSlotParameters(impBuilder, impExt.getBidder().getAdSlot(), banner);
+        if (banner != null) {
+            impBuilder.banner(assignSizesIfMissing(impBuilder.build().getBanner()));
```

Comment:
Pls take a look on this line, can this be used or updates should be prepared

Reply:
Yeah, it looks kind of cryptic.
Can we do `assignSizesIfMissing` actions in `updateImpWithAdSlotParameters` ?